Don't process xthin blocks unless they extend the chain

kyuupichan opened the pull request:
Also:

- cleans up validation logic; most is now done in AcceptBlockHeader
- reworks the logic so cs_main is held only once

Commit:
Don't process xthin blocks unless they extend the chain

Also:

- cleans up validation logic; most is now done in AcceptBlockHeader
- reworks the logic so cs_main is held only once

sickpig commented:
failure is related to formatting of `thinblock.cpp` 

https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/230275641#L1184

it's not about actual tests failing

Commit:
Fix formatting, tweak logging.

When the xthinblock is rejected for not extending the chain, request
a full block rather than a thin block - our current mempool is unlikely
to help the thinblock code much for an old block.

sickpig commented:
@kyuupichan compiled locally, all tets passed, tested on mainnet for more than day now and it didn't show any problem

Commit:
Add missing return

gandrewstone committed:
add updateblockavailability, and fix return false vs true

ptschip commented on src/thinblock.cpp:
```diff
@@ -261,84 +261,90 @@ bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strComm
+        CValidationState state;
+        CBlockIndex *pIndex = NULL;
+        if (!AcceptBlockHeader(thinBlock.header, state, Params(), &pIndex))
+        {
```
if we get an out of order thinblock somehow won't we end up invalidating this header and then the chain will  no longer extend?

ptschip commented on src/thinblock.cpp:
```diff
@@ -312,34 +311,34 @@ bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strComm
             vector<CInv> vGetData;
-            vGetData.push_back(CInv(MSG_THINBLOCK, inv.hash));
-
+            vGetData.push_back(inv);
```
i believe this is incorrect, you've formatted previously a inv with no hash...you have to give it the block hash.  i would put here the whole string so there is no confusion

`+ vGetData.push_back(CInv(MSG_BLOCK, inv.hash));`

ptschip reviewed (approved):
i havn't tested it but on further review the code looks good.

gandrewstone committed:
merge

Commit:
Merge pull request #3 from gandrewstone/xthinchecks

Xthinchecks
